source/event_stream_rpc_client.c

```diff
@@ -785,12 +788,15 @@ static void s_route_message_by_type(
                 (void *)connection,
                 (void *)continuation);
             aws_atomic_store_int(&continuation->is_closed, 1U);
+            int was_present = 0;
             aws_mutex_lock(&connection->stream_lock);
-            aws_hash_table_remove(&connection->continuation_table, &stream_id, NULL, NULL);
+            aws_hash_table_remove(&connection->continuation_table, &stream_id, NULL, &was_present);
             aws_mutex_unlock(&connection->stream_lock);
 
             /* Note that we do not invoke callback while holding lock */
-            s_complete_continuation(continuation);
+            if (was_present) {
+                s_complete_continuation(continuation);
```

Comment:
This improves things, but I'm still worried about `s_clear_continuation_table()` (which is called from `aws_event_stream_rpc_client_connection_close()` and `s_on_channel_shutdown_fn()`)

It seems to mark ALL continuations closed, and fire their callbacks, but does NOT remove them from the hashtable. I'm haven't dug through every possible execution path, but this seems fragile, like we could do the callback/release but leave them in the table, then callback/release again from one of these places you just fixed

Reply:
I agree